pkg/psalabelsyncer/sccrolecache.go

```diff
@@ -0,0 +1,440 @@
+// at most one of role/clusterrole is always non-nil
+type roleBindingObj struct {
+	roleBinding        *rbacv1.RoleBinding
+	clusterRoleBinding *rbacv1.ClusterRoleBinding
+}
+
+func newRoleBindingObj(obj interface{}) (*roleBindingObj, error) {
+	if o, ok := obj.(*rbacv1.ClusterRoleBinding); ok {
+		return &roleBindingObj{
+			clusterRoleBinding: o,
+		}, nil
+	} else if o, ok := obj.(*rbacv1.RoleBinding); ok {
+		return &roleBindingObj{
+			roleBinding: o,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("the object is neither a RoleBinding, nor a ClusterRoleBinding: %T", obj)
+}
+
+func (r *roleBindingObj) RoleRef() rbacv1.RoleRef {
+	if binding := r.clusterRoleBinding; binding != nil {
+		return binding.RoleRef
+	}
+	return r.roleBinding.RoleRef
+}
+
+func (r *roleBindingObj) Subjects() []rbacv1.Subject {
+	if binding := r.clusterRoleBinding; binding != nil {
+		return binding.Subjects
+	}
+	return r.roleBinding.Subjects
+}
+
+func (r *roleBindingObj) Namespace() string {
+	if r.clusterRoleBinding != nil {
+		return ""
+	}
+	return r.roleBinding.Namespace
+}
+
+// AppliesToNS returns true if:
+// - the wrapped object is a cluster role binding
+// - the namespace of a wrapped rolebinding matches the namespace supplied in `ns`
+func (r *roleBindingObj) AppliesToNS(ns string) bool {
+	if r.clusterRoleBinding != nil {
+		return true
+	}
+	return ns == r.roleBinding.Namespace
+}
+
+// roleObj helps to handle roles and clusterroles in a generic manner
+type roleObj struct {
+	role        *rbacv1.Role
+	clusterRole *rbacv1.ClusterRole
+}
+
+func newRoleObj(obj interface{}) (*roleObj, error) {
+	switch r := obj.(type) {
+	case *rbacv1.ClusterRole:
+		return &roleObj{
+			clusterRole: r,
+		}, nil
+	case *rbacv1.Role:
+		return &roleObj{
+			role: r,
+		}, nil
+	case *roleObj:
+		return r, nil
+	default:
+		return nil, fmt.Errorf("the object is neither a Role, nor a ClusterRole: %T", obj)
+	}
+}
+
+func (r *roleObj) Rules() []rbacv1.PolicyRule {
+	if role := r.clusterRole; role != nil {
+		return role.Rules
+	}
+	return r.role.Rules
+}
+
+func (r *roleObj) Name() string {
+	if role := r.clusterRole; role != nil {
+		return role.Name
+	}
+	return r.role.Name
+}
+
+func (r *roleObj) Namespace() string {
+	if role := r.clusterRole; role != nil {
+		return role.Namespace
+	}
+	return r.role.Namespace
+}
+
+// BySAIndexKeys is a cache.IndexFunc indexing function that shall be used on
+// rolebinding and clusterrolebinding informer caches.
+// It retrieves the subjects of the incoming object and if there are SA, SA groups
+// or the system:authenticated group subjects, these will all be returned as a slice
+// of strings to create an index for the SA or SA group.
+func BySAIndexKeys(obj interface{}) ([]string, error) {
+	roleBinding, err := newRoleBindingObj(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceAccounts := []string{}
+	for _, subject := range roleBinding.Subjects() {
+		if subject.APIGroup == "" && subject.Kind == "ServiceAccount" {
```

Comment:
minor: I prefer switch/case and summarizing someting like

```go
IsGroupBinding := subject.APIGroup == rbacv1.GroupName && subject.Kind == "Group"
isAllServiceAccount := subject.Name == serviceaccount.AllServiceAccountsGroup
isNamespacedServiceAccountGroup := strings.HasPrefix(subject.Name, serviceaccount.ServiceAccountGroupPrefix))
```

Reply:
I tried, it turns out to be quite a lot of string comparisons that may be unnecessary if I wanted to keep summaries like that. I can keep a switch-case for a) direct SAs; b) groups but the rest I'd keep the same.